src/Database/LSMTree/Internal/PageAcc.hs

```diff
@@ -317,8 +317,10 @@ serializePageAccN size MPageAcc {..} = do
             | otherwise = do
                   -- traceM $ "loop " ++ show (i, ko, kd, vo, vd)
 
-                  SerialisedKey   (RawBytes (PV.Vector koff klen kba)) <- MV.read mpKeys i
-                  SerialisedValue (RawBytes (PV.Vector voff vlen vba)) <- MV.read mpValues i
+                  k <- MV.read mpKeys i
+                  v <- MV.read mpValues i
+                  let SerialisedKey   (RawBytes (PV.Vector koff klen kba)) = k
```

Comment:
why this change?

Is having lazy pattern match important?

Reply:
Unfortunately having a pattern synonym means the pattern match is always considered partial for the purpose of `MonadFail`. So if one uses the pattern here in the monad bind then we get complaints about not having a fail constraint. Using COMPLETE pragmas doesn't help apparently.

We can make those pattern matches strict.

Reply:
Ah, so this https://gitlab.haskell.org/ghc/ghc/-/issues/15681

I kindly ask you to mention that you are hitting that as well. Maybe at some point, GHC HQ acknowledges that that issue is really should be fixed even if you cannot get ICFP paper out of that.